File: algorithms/sort/mergeSort.go
package sort

func MSort(arr []int) []int {
	// 合并排序，用的也是分而治之
	if len(arr) <= 1 {
		return arr
	}
	num := len(arr) / 2
	return merge(MSort(arr[:num]), MSort(arr[num:]))
}
func merge(left, right []int) []int {
	var i, j int = 0, 0
	result := []int{}
	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}

	result = append(result, left[i:]...)
	result = append(result, right[j:]...)
	return result
}

File: sort_dart/mergeSort.dart
merge(List left, List right) {
  List res = [];

  int i = 0;
  int j = 0;
  while (i < left.length && j < right.length) {
    if (left[i] < right[j]) {
      res.add(left[i]);
      i++;
    } else {
      res.add(right[j]);
      j++;
    }
  }
  res.addAll(left.sublist(i));
  res.addAll(right.sublist(j));
  return res;
}

mergeSort(List arr) {
  if (arr.length < 2) {
    return arr;
  }
  int mid = arr.length ~/ 2;
  List left = arr.sublist(0, mid);
  List right = arr.sublist(mid);
  return merge(mergeSort(left), mergeSort(right));
}

main(List<String> args) {
  var arr = [2, 6, 5, 4, 8, 0, 9, 1, 2, 2];
}
